bob-apps/bin/bobc.rs

```diff
@@ -1,137 +1,472 @@
+        }
+    }
+}
+
+const KEY_ARG: &str = "key";
+const KEY_SIZE_ARG: &str = "key-size";
+const HOST_ARG: &str = "host";
+const PORT_ARG: &str = "port";
+const FILE_ARG: &str = "file";
+
+const PUT_SC: &str = "put";
+const GET_SC: &str = "get";
+const EXIST_SC: &str = "exist";
+
+#[derive(Debug)]
+enum ParseError {
+    FilePattern,
+    KeyPattern,
+}
+
+#[derive(Debug)]
+struct AppArgs {
+    subcommand: String,
+    file_pattern: Option<FilePattern>,
+    key_pattern: Option<KeyPattern>,
+    keysize: usize,
+    host: String,
+    port: u16,
+}
+
+impl AppArgs {
+    fn from_matches(matches: ArgMatches) -> Self {
+        let (sc, sub_matches) = matches.subcommand();
+        let sub_matches = sub_matches.unwrap();
+        let file_pattern = match sub_matches.value_of(FILE_ARG) {
+            None => None,
+            Some(p) => match Self::parse_file_pattern(p) {
+                Ok(val) => Some(val),
+                Err(e) => panic!("{:?}", e), // should send an error to the caller
+            },
+        };
+        let key_pattern = match sub_matches.value_of(KEY_ARG) {
+            None => None,
+            Some(p) => match Self::parse_key_pattern(p) {
+                Ok(val) => Some(val),
+                Err(e) => panic!("{:?}", e), // again
+            },
+        };
+
+        Self {
+            subcommand: sc.to_string(),
+            file_pattern,
+            key_pattern,
+            keysize: sub_matches.value_or_default(KEY_SIZE_ARG),
+            host: sub_matches.value_or_default(HOST_ARG),
+            port: sub_matches.value_or_default(PORT_ARG),
+        }
+    }
+    fn parse_file_pattern(file_pattern: &str) -> Result<FilePattern, ParseError> {
+        let path = PathBuf::from(file_pattern);
+        let name = path.file_name().unwrap().to_str().unwrap();
+
+        let re = Regex::new("\\{key\\}").unwrap();
+
+        match re.find_iter(name).count() {
+            0 => Ok(FilePattern::WithoutRE(file_pattern.to_string())),
+            1 => {
+                let path = PathBuf::from(file_pattern);
+
+                let dir = match path.parent().unwrap() {
+                    p if p == Path::new("") => Path::new("."),
+                    p => p,
+                };
+                Ok(FilePattern::WithRE(
+                    RePattern::new(file_pattern),
+                    dir.to_owned(),
+                ))
+            }
+            _ => Err(ParseError::FilePattern),
+        }
+    }
+    fn parse_key_pattern(key_arg_cont: &str) -> Result<KeyPattern, ParseError> {
+        if let Ok(key) = key_arg_cont.parse() {
+            return Ok(KeyPattern::Single(key));
+        }
+        if key_arg_cont.contains("-") {
+            let v: Vec<&str> = key_arg_cont.split("-").collect();
+            if v.len() != 2 {
+                return Err(ParseError::KeyPattern);
+            }
+            let v0 = v[0].parse().unwrap();
+            let v1 = v[1].parse().unwrap();
```

Comment:
If parse failed, `unwrap` will panic. You should return `Err(ParseError::KeyPattern)` instead